### PR TITLE
Use GitLab-CI

### PR DESCRIPTION
**Pull request by Tomahna**
Would something like that do the trick for #1922 ?

I can't seem to access https://gitlab.com/ensime/ensime-server/, are you sure the repository is public ?

Anyway I've created an integration to test, you can find it here https://gitlab.com/krauscher/ensime-server.

### CODE REVIEW

**Comment by fommil** [comment]
nice, almost there. The caching strategy I use in scalaz-deriving should work better than this... most of what you're caching is already in the base image but the gitlab cache is better served by doing the COURSIER downloads 

https://gitlab.com/fommil/scalaz-deriving/blob/master/.gitlab-ci.yml

I *think* that works for merges doing the cache updates, but I am not 100% sure about it.

One big drawback of gitlabci is that it is really hard to setup a matrix build, so if we want to add the scala 2.11 build back in, it is going to be a real pain.

**Comment by Tomahna** [comment]
Yeah, you're right, I didn't check your docker image

Well, there is no matrix build in GitLab. However, we do not need to run the entire pipeline against 2.11 and 2.12. test and it-test should be enough so I can just add the tasks manually.

**Comment by Tomahna** [comment]
A few additional thoughts:
- I don't think there is any direct cache merge, the cache is just pulled at the beginning and pushed at the end (with new files). It seems fine as it means the cache will just converge to its final content.
- GitLab public shared runners are small (Digital Ocean 4GB 2vCPUs droplets) but autoscaling. Thus, it's better to have many smaller tasks than a few big ones.

So it should run better now, the first stage takes between 10 to 15 minutes to complete. The integration tests takes around 50 minutes. It's not 100% reliable though.

**Comment by fommil** [comment]
in my experience the cache does not converge which just means the machine can't cache the cache and every run has to upload a huge cache to S3, which has the end result of being slower than not using the cache (plus costing me more money when using my own runner and S3). Hence why I only push the cache on merges to master and pull it on PRs.

**Comment by fommil** [comment]
ouch, 50 minutes is pretty harsh. Maybe we should time travis-ci for GNU/Linux to see if it's as bad.

**Comment by Tomahna** [comment]
I don't think I see your problem here. If you just cache coursier dependencies, the worst thing that might happen is you cache dependencies that are not yet or not anymore in use. You can solve it by invalidating the cache once in a while.

Concerning the integration tests, I might be missing a tweak here and there, but I've got a lot of OOM errors when trying to be faster.

Anyway, I agree we should probably test if travis behaves better.

**Comment by fommil** [comment]
@Tomahna the problem is that the cache gets downloaded and uploaded to S3 on every single PR. It never converges. I don't know why.

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -6,7 +6,7 @@ cache:
 variables:
   AKKA_TEST_TIMEFACTOR: 10
   COURSIER_CACHE: "${CI_PROJECT_DIR}/project/.sbt/coursier"
-  SBT_TASK_LIMIT: 4
+  SBT_OPTS: "-XX:+CMSClassUnloadingEnabled  -XX:+UseConcMarkSweepGC -Xss2m -Xms3g -Xmx3g"
```
Comment (fommil) [inline .gitlab-ci.yml]: no need, we have a `.jvmopts` file

**Comment by Tomahna** [comment]
Yeah i saw it, it's mostly a test at that point.

File: .gitlab-ci.yml (hunk 1 above)
Comment (Tomahna) [reply .gitlab-ci.yml]: Yeah i saw it, it's mostly a test at that point.

**Pull request merged**